### ISSUE TITLE
OAuth2

### ISSUE THREAD

**Issue by alantweir** [otris/ews-cpp#148]
Microsoft is dropping Basic authentication on October 13th, 2020 and will only be supporting OAuth2 for authenticating with O365 after that date. [Link](https://blogs.technet.microsoft.com/exchange/2018/07/03/upcoming-changes-to-exchange-web-services-ews-api-for-office-365/ ) 

Can OAuth2 be added prior  to October 2020 to allow time for testing?

Alan

**Comment by ZaMaZaN4iK** [comment]
I agree. It's very importatnt even for now because of OAuth2 is the most popular authentication way nowadays.

**Comment by kmorrissey-mersive** [comment]
We currently use this library to access O365, so we also need some O365-OAuth2 solution.

**Comment by kmorrissey-mersive** [comment]
My company is likely able to provide resources to help solve this. If anyone else can join in, please leave a comment here.

**Comment by ZaMaZaN4iK** [comment]
I am also interested in this. @kmorrissey-mersive about which kind of resources are you speaking?

**Comment by kmorrissey-mersive** [comment]
I'm fairly sure my company will be willing to assign a developer to this problem at least part-time (possibly me), more staffing if needed. We use this library in a product to communicate with EWS, especially to O365. I don't know of a replacement for it, so I suspect we will be committed to working on this. We'll need to propose an interface to isolate ews-cpp from a specific OAuth implementation as much as practical.

**Comment by ZaMaZaN4iK** [comment]
As mentioned earlier in https://github.com/otris/ews-cpp/issues/146 we need first to implement Azure AD Auth library, then integrate to ews-cpp

**Comment by kmorrissey-mersive** [comment]
Is anyone in a position to judge whether https://github.com/Microsoft/cpprestsdk has code applicable to what is needed?

**Comment by idolum** [comment]
@kmorrissey-mersive I you can assing a developer this would be great! Unfortunately, we had to reprioritize our own task items giving us a challenge to work in this issue right now. So, any help is more than welcome and provide any support possible!

Regarding the cpprestdsk: Looks promissing, but I for my self have to take a close look, before I can give a propert answer.

**Comment by idolum** [comment]
I discussed this issue with a colleague how this can be implemented. The first rough ideas are

* Add another credentials class `oauth2_credentials`
* Uses the [Grant Type](https://oauth.net/2/grant-types/) "Password" to get an access token
* The first invocation of `certify()` retrieves the access token used then in any further `certify()` invocation

But these are our first step in this direction to get a better undetstanding how this can work. But any remark or other ideas are much appreciated.

**Comment by kmorrissey-mersive** [comment]
We will be using this with an embedded device that talks to Exchange but is administered remotely, so it's important that ews-cpp itself does not get involved with user interaction. If I understand right, oauth2_credentials should just have an authorization (right word?) token and a refresh token. Ews-cpp can use the refresh token if needed, but it should not try to trigger an interaction if the refresh fails.

**Comment by ZaMaZaN4iK** [comment]
@kmorrissey-mersive yes - you understand correctly :)

**Comment by ZaMaZaN4iK** [comment]
@idolum sounds reasonable. Btw, you should keep in mind that we also should be able to work with JWT. So we can use some existent library for JWT (preferred way) or implement some class from scratch in the library (but I don't like NIH solutions).

**Comment by kmorrissey-mersive** [comment]
Perhaps we can define some sort of wrapper class that ews-cpp uses that can be implemented in separate headers to work with different OAuth libraries. I'm all for keeping ews-cpp OAuth library neutral.

**Comment by idolum** [comment]
We figured out that the Password grant type restricted in v2 of the Authorization protocol to [Azure AD tenants](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth-ropc) and does not support personal accounts. Thus, the [Client Credentials](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-client-creds-grant-flow) seems more logical. If we understood it right, this grant type is more appropriate for embedded device. But the [Device Code](https://oauth.net/2/grant-types/device-code/) might be also a candidate here.

@ZaMaZaN4iK By JWT you mean JSON Web Tokens, right?

@kmorrissey-mersive What do you mean by library? As I see it right, no specific library is needed. Just one for sending and receiving HTTP requests. But the implementation should be neutral regarding OAuth service implementations, so that "talk" to the OAuth implementation of Azure as well as any other OAuth implementation.

**Comment by ZaMaZaN4iK** [comment]
> By JWT you mean JSON Web Tokens, right?

Right

**Comment by idolum** [comment]
As you may have notes @rbns has opend a pull request #154 with the first draft of an OAuth2 implementation. There is still a clang issue, but not related to the implementation itself. Any remarks is much appreciated!

Does anyone of you have the chance to test this somehow with an OAuth2 service and/or Azure AD already in place? Right now we haven't a proper test environment available and need to set it up from scratch.

**Comment by idolum** [comment]
Some issues that needs to be addressed:

* `get_exchange_web_services_url()` needs to adapted to work with the `oauth2_client_credentials`
* `expires_on` should be `expires_in`
* Azure/Office 365 requires also a `scope` for requesting the credentials
* Access token is not provided properly in the requests to Office 365

But these issues will be addressed in a follwing pull request.

**Comment by idolum** [comment]
Fixed the issues mentioned above in #163 
Feel free to review the changes.

Added also an example for accessing Office 365 with OAuth2. But until could not make it work. I get a 401 with the header message

`x-ms-diagnostics: 2000008;reason="The token contains no permissions, or permissions can not be understood.";error_category="invalid_grant"`

Any idea? Does anybody know, which permissions the client should be granted to and what the right scope is? I used `https://outlook.office.com/.default`.

**Comment by ZaMaZaN4iK** [comment]
@idolum try this for EWS: `https://outlook.office365.com/.default` or `https://outlook.office365.com/ews/Exchange.asmx/.default`

**Comment by ZaMaZaN4iK** [comment]
About authentication and other such stuff: you can mention me here. I've already have implemented OAuth authentication for MS GRaph API, EWS API and SharePoint Online REST API. But I cannot open the code because that's from my work :( (but I'll try to help as much as possible).

**Comment by idolum** [comment]
Thanks @ZaMaZaN4iK 

Tried you proposals, but did not succeed either:

* `https://outlook.office365.com/.default`: I get the same error as above.
* `https://outlook.office365.com/ews/Exchange.asmx/.default`: I get an error as well but a different one when the access token is requested:

`AADSTS500011: The resource principal named https://outlook.office365.com/ews/Exchange.asmx was not found in the tenant named <removed>.onmicrosoft.com. This can happen if the application has not been installed by the administrator of the tenant or consented to by any user in the tenant. You might have sent your authentication request to the wrong tenant.`

At the moment, I don't think that this is an issue with the implemenation, but rather with the configuration. But according the granted permission, the client should be able to access the Exchange via EWS.

**Comment by ZaMaZaN4iK** [comment]
Yeah, seems like missed configuration on the tenant. Did you accept your application on a tenant? Which permissions are selected for your application?

**Comment by idolum** [comment]
I think so. Unfortunatly, it is a tenant of a customer and I don't have direct access. But I will ask. The following permissions are granted to the client:

* EWS.AccessAsUser.All (Exchange and Graph)
* User.Read (Exchange)

I wonder, if this is related to the client_credentials flow and some kind of impersonation has to be done as well.

**Comment by ZaMaZaN4iK** [comment]
> kind of impersonation has to be done as well.

Yes. Impersonation has to be done in EWS. Without this it doesn't work even with application.

**Comment by idolum** [comment]
I had a conference call with an EWS and we solved the issue. A proper permission was missing on the server side. The client needs the application permission "FullAccess" to get access via EWS. After adding the impersonation to the example, authentification with OAuth2 works. So, from this perspective OAuth2 works.

But we identified some use cases where this is not appropriate. It seems for such use cases we have to support the ROPC flow as well:
https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth-ropc
